### PR TITLE
Prepare for creating models with SSH keys.

### PR DESCRIPTION
**Pull request by frankban**
The deployment flow component now has the ability to store an SSH key in its state, and to use that when deploying the workload and creating a new model.

### CODE REVIEW

**Comment by frankban** [comment]
QA: check that creating a new model from the GUI still works.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6616/

**Comment by Lukewh** [comment]
Creating new model from GUI still works

File: jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js [hunk 1]
```diff
@@ -177,6 +178,16 @@ YUI.add('deployment-flow', function() {
     },
 
     /**
+      Store the provided SSH key in state.
+
+      @method _setSSHKey
+      @param {String} key The SSH key.
+    */
+    _setSSHKey: function(key) {
+      this.setState({sshKey: key});
```
Comment (mitechie) [inline jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js]: can this be more than one key?

Comment (mitechie) [inline jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js]: Just hit a thought on should we be looking to leverage ssh keys in LP or GH APIs for any of this?

File: jujugui/static/gui/src/app/components/deployment-flow/test-deployment-flow.js [hunk 2]
```diff
@@ -818,13 +821,69 @@ describe('DeploymentFlow', function() {
...
+      cloud: 'cloud',
+      region: 'north'
+    });
     assert.equal(changeState.callCount, 1);
   });
 
+  it('can deploy with SSH keys', function() {
+    var deploy = sinon.stub().callsArg(0);
+    var changeState = sinon.stub();
+    var renderer = jsTestUtils.shallowRender(
```
Comment (mitechie) [inline jujugui/static/gui/src/app/components/deployment-flow/test-deployment-flow.js]: wow...

File: jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js [hunk 3]
```diff
@@ -250,15 +261,19 @@ YUI.add('deployment-flow', function() {
...
-        this._handleClose, true, modelName, credential, cloud, region);
+      const args = {
+        credential: this.state.credential,
+        cloud: this.state.cloud && this.state.cloud.name || undefined,
+        region: this.state.region
+      };
+      if (this.state.sshKey) {
+        args.config = {'authorized-keys': this.state.sshKey};
+      }
+      this.props.deploy(this._handleClose, true, modelName, args);
```
Comment (mitechie) [inline jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js]: the api for creating a new model is deploy? Is that Juju's API for it or our own?

File: jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js (hunk 1 above)
Comment (frankban) [inline jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js]: It does not seem that we can pass more than one key when creating the model.
LP or github API is a good idea, but it does not affect this branch. Here we are only setting up a system to store the key, where does that key come from is another question. Initially we'll just support copy/pasting a key. Later, we can improve that as you suggested.

File: jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js (hunk 3 above)
Comment (frankban) [inline jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js]: It's our own wrapper, defined in views.utils, and yes, poorly named.

File: jujugui/static/gui/src/app/components/deployment-flow/test-deployment-flow.js (hunk 2 above)
Comment (frankban) [inline jujugui/static/gui/src/app/components/deployment-flow/test-deployment-flow.js]: yeah.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6618/

**Comment by frankban** [comment]
Thanks for the reviews!
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**